saasherder/changelog.py

```diff
@@ -1,140 +1,186 @@
+    def render(self, header_template, section_template, commit_template):
+        """Pretty print the change log"""
 
-        return self.run(cmd)
+        # Build the body
+        header = header_template.format(url=self.url, old=self.old, new=self.new)
 
-    def checkout(self, service, version):
-        """Get a service checked out at specified version"""
+        body = [header]
 
-        name = service['name']
-        if os.path.exists(self.worktree(name)):
-            logger.info("Found git worktree for {}; updating".format(name))
-            self._fetch(service)
-        else:
-            logger.info("Cloning git worktree for {}".format(name))
-            self._clone(service)
+        for service in self.changelog.changed_services:
+            names_list = "\n".join(["- {}".format(n) for n in service['names']])
+            url = service['url']
 
-        worktree = self.worktree(service['name'])
-        cmd = "cd {} && git checkout {} ".format(worktree, version)
+            section = section_template.format(name=service['name'],
+                                              names_list=names_list,
+                                              url=url,
+                                              old=service['old'],
+                                              new=service['new'],
+                                              log=self.log(service, commit_template),
+                                              last_changed=self.last_changed(service))
+            body.append(section)
 
-        return self.run(cmd)
+        return '\n'.join(body)
 
-    def log(self, service, old, new):
+    def log(self, service, commit_template):
         """Get the changelog for one service"""
+        template = commit_template.format(url=service['url'])
+        git_log_cmd = "git log --format='{}' {}...{}".format(template, service['old'], service['new'])
+        return self.changelog.service_run(service, git_log_cmd)
 
-        worktree = self.worktree(service['name'])
-        remote = service['url'].strip('/')
-        template = '[%h](_REMOTE_/commit/%H) %s'
+    def last_changed(self, service):
+        """Get the commit time for service at specific commit"""
+        return self.changelog.service_run(service, "git log -1 --format=%cd {}".format(service['new'])).strip()
 
-        cmd = ("cd {} && git log --format='{}' {}...{}"
-               .format(worktree, template, old, new))
+class Changelog(object):
 
-        return self.run(cmd).replace('_REMOTE_', remote)
+    workspace = "_workspace/"
 
-    def last_changed(self, service, commit):
-        """Get the commit time for service at specific commit"""
+    def __init__(self, parent):
+        self.parent = parent
 
-        worktree = self.worktree(service['name'])
-        cmd = "cd {} && git log -1 --format=%cd {}".format(worktree, commit)
+        # Key: url of the service
+        # Value: list of services that have changed
+        self.changed_services = []
```

Comment:
You are using an array and a hash with the same name. Could you simplify this a bit maybe?

Reply:
Agreed.